http/src/main/scala/org/http4s/blaze/http/ClientSessionManagerImpl.scala

```
@@ -0,0 +1,167 @@
+
+          case h2: Http2ClientSession if h2.quality > 0.1 =>
+            session = h2
+
+          case _: Http2ClientSession => () // nop
+
+          case h1: Http1ClientSession =>
+            it.remove()
+            if (h1.status != Closed) { // Should never be busy
+              session = h1 // found a suitable HTTP/1.x session.
+            }
+        }
+
+        // if we took the last session, drop the collection
+        if (sessions.isEmpty) {
+          sessionCache.remove(id)
+        }
+
+        session
+    }
+  }
+
+  private[this] def createNewSession(urlComposition: UrlComposition, id: ConnectionId): Future[HttpClientSession] = {
+    logger.debug(s"Creating new session for id $id")
+    val p = Promise[HttpClientSession]
+
+    socketFactory.connect(urlComposition.getAddress).onComplete {
+      case Failure(e) => p.tryFailure(e)
+      case Success(head) =>
+        val clientStage = new Http1ClientStage(config)
+        var builder = LeafBuilder(clientStage)
+        if (urlComposition.scheme == "https") {
+          val engine = config.getClientSslEngine()
+          engine.setUseClientMode(true)
+          builder = builder.prepend(new SSLStage(engine))
+        }
+
+        builder.base(head)
+        head.sendInboundCommand(Command.Connected)
+        p.trySuccess(new Http1SessionProxy(id, clientStage))
+      }
+
+    p.future
+  }
+
+  /** Return the session to the pool.
+    *
+    * Depending on the state of the session and the nature of the pool, this may
+    * either cache the session for future use or close it.
+    */
+  override def returnSession(session: HttpClientSession): Unit = {
+    logger.debug(s"Returning session $session")
+    session match {
+      case _ if session.isClosed => () // nop
+      case _: Http2ClientSession => () // nop
+      case h1: Http1ClientSession if !h1.isReady =>
+        logger.debug(s"Closing unready session $h1")
+        h1.closeNow() // we just orphan the Future. Don't care.
```

Are you throwing away an exception if it fails?

---

Yes, I didn't think it would be valuable. I can log on the failure pathway if you like: could be interesting to know.

---

Turning `Future[Unit]` into `Unit` is essentially an empty catch block.  I have been burned rarely, but badly, by this.  If you're going to discard it, at least log the failures.

(I wouldn't have noticed it if not for your comment.)